### ISSUE TITLE
Одновременное создание разных инстансов фичи

### ISSUE THREAD

**Issue by arttttt** [PavelSidyakin/WeatherForecast#1]
Исходя из текущей реализации, ComponentHolder может держать только один инстанс фичи, соответственно, возникает вопрос: как разрулить ситуацию, когда требуются независимые инстансы фичи, плюс, чтобы у них были разные зависимости?

**Comment by PavelSidyakin** [comment]
Это не проблема способа склейки компонентов (ComponentHolder и других вспомогательных классов). Это скорее относится к логике самой фичи. Вообще, не следует как-то рассчитывать на то, сколько будет жить компонент и какой у него lifecycle.

Если фича предполагает существование нескольких разных инстансов какой-либо сущности, то из фичи должна торчать **фабрика** этой сущности. 

Если ещё и разные зависимости нужно использовать, то тут, вижу два варианта: 

1. Все возможные зависимости прокинуть в фичу и в вышеупомянутой фабрике выбирать. 
2. Сделать отдельные фичи. Тогда их легко будет инициализировать разными зависимостями.

На выбор того или другого варианта может повлиять насколько разный функционал хочется получить. Если разница в функционале незначительная, то проще использовать вариант 1). 
Если функционал сильно будет отличаться, то это уже скорее вариант 2). Плюс потом будет проще разрабатывать фичи отдельно. Если при этом есть общий функционал, то его можно вынести в отдельный модуль и прокинуть зависимостями.

**Comment by arttttt** [comment]
Все равно не особо понятно. Например, у нас есть две вкладки нижней навигации. Есть фича(FeatureApi), которая представляет из себя некий флоу из нескольких экранов, данная фича принимает как dependency некий слушатель и умеет возвращать(из FeatureApi) что-то, например, через Observable<T>, не суть.

Каждая вкладка может запустить флоу целой фичи одновременно.

Получается, что в таком случае оптимальным вариантом будет возвращать из фичи(FeatureApi) фабрику создания самой себя?

**Comment by PavelSidyakin** [comment]
> Каждая вкладка может запустить флоу целой фичи одновременно.

Т.е. из фичи торчит фрагмент? Вкладка запрашивает у фичи фрагмент?

> Получается, что в таком случае оптимальным вариантом будет возвращать из фичи(FeatureApi) фабрику создания самой себя?

Нет, не надо так делать. 
Я имел в ввиду, что должна быть фабрика той сущности, для которой нужны разные инстансы. Т.е., например, из FeatureApi может торчать интерфейс MyInteractorFactory, который возвращает инстансы MyInteractor'а. Либо фабрика фрагментов, которая возвращает разные инстансы фрагмента. В фабрику можно передать параметры, чтобы создать сущности с разным функционалом. 
При этом инстанс компонента (фичи) один. Но оно может отдавать разные инстансы каких-то других сущностей.

**Comment by arttttt** [comment]
> Т.е. из фичи торчит фрагмент? Вкладка запрашивает у фичи фрагмент?

На самом деле не принципиально, просто для примера одновременного использования их двух мест.

У меня сейчас проблема в том, что в FeatureDependencies нужно передать реализацию интерфейса, которая будет будет разной и зависит от места, из которого фича была вызвана

**Comment by PavelSidyakin** [comment]
> передать реализацию интерфейса, которая будет разной и зависит от места, из которого фича была вызвана

Скорее всего, тогда это не должно быть в Dependencies фичи. Этот интерфейс тогда представляет собой как-бы конфиг, и должен передаваться в фабрику, которая торчит из FeatureApi. И уже фабрика создаст экземпляр сущности (не фичи!) с нужным функционалом.

**Comment by arttttt** [comment]
Привет, решил накидать небольшой пример проблемы, которую описывал: https://github.com/arttttt/ArchSample/tree/dependencies_issue

Суть такая:

Есть фрагменты: Tab1, Tab2 и Shared. Соответственно, Shared фрагмент может быть вызван как из Tab1, так и из Tab2, причем одновременно.
Shared фрагмент требует для работы предоставление интерфейса BackPressedHandler. Tab1 и Tab2 имеют свои реализации интерфейса BackPressedHandler. Соответственно, нужно как-то в рантайме разрулить зависимости.

Как я понимаю, подобная фабрика фрагмента тут работать не будет, потому что BackPressedHandler потеряется после восстановления состояния. Соответственно, не могли бы помочь найти наилучшее решение подобного кейса.

```
fun interface SharedScreenFragmentFactory {
    fun create(backPressedHandler: BackPressedHandler): Fragment
}
```

**Comment by PavelSidyakin** [comment]
Привет! Сорян за долгий ответ, я в отпуске.
Коллбэк не надо выносить в зависимости фичи. 
В частности с фрагментом, я бы не делал фабрику для него, а просто позволил ему торчать из фичи. Да, знаю, что в примере фабрика, но по опыту с фрагментами, его проще просто сделать доступным напрямую. Можно это записать в ограничения системы. Для других типов фич (не связанных с UI Android), такие послабления не допустимы.
В этом случае можно коллбек задавать напрямую фрагменту через ViewModel и LiveData. И восстанавливать его при пересоздании parent.
Пример можно глянуть тут: https://github.com/PavelSidyakin/ProductList 
См. задание коллбэка в ProductListFragment.kt

**Comment by arttttt** [comment]
```
В этом случае можно коллбек задавать напрямую фрагменту через ViewModel и LiveData
```
Не нравятся они мне, да и не используем в проекте

**Comment by PavelSidyakin** [comment]
Можно тогда просто коллбэк задавать фрагменту и вешать его в onAttachFragment() у parent. Но этот способ не будет работать с nav component.

**Comment by arttttt** [comment]
Да, не особо удобно получается

> Но этот способ не будет работать с nav component.

будет, просто придется дважды ставить addFragmentOnAttachListener:
первый повесить на parent, чтобы найти NavHostFragment 
второй повесить на NavHostFragment чтобы найти требуемый фрагмент

**Comment by PavelSidyakin** [comment]
Да, не очень красиво. Но.. это андроид. Для модулей, предоставляющих платформенные вещи - приходится идти на компромиссы. 
В любом случае - это не проблема склейки модулей. Интерфейс dependencies - это для "постоянных" зависимостей (интеракторы, репозитории и т.п.). Коллбэк там явно не в тему.
